### ISSUE TITLE
TF2 Grow Scores calculation

### ISSUE THREAD

**Issue by AmoghDabholkar** [google-research/rigl#14]
The TF1 code repo calculates grow scores based on the gradients of masked variables (after multiplication of masks and vars) 
Refer: https://github.com/google-research/rigl/blob/0f029735f84e0120df05512244510e8ed48a4461/rigl/sparse_optimizers_base.py#L481
where masked_weights are fetched from https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/model_pruning/python/pruning.py#L258

whereas in the TF2 code, there is no such thing, which I think affects the performance of RigL in general. Is there any way to apply a fix for this? 
Also, is this because if the TF2 code the updates are made using a part of validation dataset? If I were to change it back to how TF1 does the updates based on the specific training batch, I would have to make the above-mentioned change right?
[We want the gradient calculation for mask update to happen after the mask and weights are multiplied and not on the original weights before parameterization as shown in the figure attached]
![image](https://user-images.githubusercontent.com/30367866/180083400-7ca8e1a9-8a8b-4fea-a424-6db42358e96c.png)

**Comment by evcu** [comment]
Indeed tensorflow model_optimization library is slightly different and not as transparent. Pruning Wrapper always applies the mask to the weights [here](https://github.com/tensorflow/model-optimization/blob/10ff67f7601cf667e5c8b9783f23a68244b62ae9/tensorflow_model_optimization/python/core/sparsity/keras/pruning_wrapper.py#L288). If I remember correctly we prune the gradients on mnist experiments, too [(see here)](https://github.com/google-research/rigl/blob/0f029735f84e0120df05512244510e8ed48a4461/rigl/rigl_tf2/train.py#L235).

So the results should be same. 

Validation set in mnist experiments are created from the training set itself (see [here](https://github.com/google-research/rigl/blob/0f029735f84e0120df05512244510e8ed48a4461/rigl/rigl_tf2/train.py#L267)). I don't think it would make much difference to use a different batch each time, but I think for simplicity it make sense to use the current batch. I think the reasoning for fixing this batch was to enable using a potentially larger batch size for updating masks.

Again I recommend using tf1 code for larger experiments.
